### PR TITLE
Fix CoffeeScript source map composition

### PR DESCRIPTION
**Pull request by klaussner**
Fixes missing CoffeeScript source maps (#8281/#8285).

* CoffeeScript has an `inlineMap` option that sets the `sourcesContent` property. However, it also includes the source map at the end of the compiled code. Since we have to process the compiled code afterwards, we can't use this option and have to set the `sourcesContent` manually. There is an issue in the CoffeeScript repo for more flexible source map options: https://github.com/jashkenas/coffeescript/issues/4342.
* The `SourceMapGenerator` was created incorrectly (its constructor doesn't accept a `SourceMapConsumer`), causing the generated source map to be empty.

A note on source map size: the produced source maps are correct but larger than necessary because of the way `applySourceMap` works. If there is no mapping from compiled CoffeeScript code to its source (`return` statements aren't mapped, for example), `applySourceMap` keeps the mappings from *app.js* to the compiled code in the source map. Therefore, the original CoffeeScript code as well as its compiled version become part of the `sourcesContent` property. The source map size could be reduced by ignoring these mappings but I think it's not worth writing a custom version of `applySourceMap` just for that.

@abernix @mjmasn @GeoffreyBooth

### CODE REVIEW

**Comment by abernix** [comment]
Awesome!  Your logic and comments make sense to me.  Probably worth linking specifically to [your comment](https://github.com/meteor/meteor/pull/8285#issuecomment-275239673) on the other issue too for others to review.

Any chance of getting some tests included on this?  I feel like we've regressed on this CoffeeScript source maps multiple times now and would like to try to harden that up.

@mjmasn, @GeoffreyBooth: Confirmation on whether this further solves the problem?

File: packages/coffeescript/plugin/compile-coffeescript.js [hunk 1]
```diff
@@ -105,10 +104,16 @@ export class CoffeeCompiler extends CachingCompiler {
       output.js = doubleRoastedCoffee.data;
 
       if (doubleRoastedCoffee.sourceMap) {
+        // Reference the compiled CoffeeScript file so that `applySourceMap`
+        // below can match it with the source map produced by the CoffeeScript
+        // compiler.
+        doubleRoastedCoffee.sourceMap.sources[0] += '.js';
```
Comment (benjamn) [inline packages/coffeescript/plugin/compile-coffeescript.js]: Maybe set `doubleRoastedCoffee.sourceMap.sources[0] = this._outputFilePath(inputFile)`, if it's the same path (I'm pretty sure it is, but haven't checked)?

Comment (klaussner) [inline packages/coffeescript/plugin/compile-coffeescript.js]: `_outputFilePath` returns the right path but without leading `/`.

**Comment by klaussner** [comment]
I think I need some help with the tests. :sweat_smile: If I understand correctly, `meteor test-packages` runs the tests in an application environment, so they won't have access to the build plugin code and the source map from `compileOneFile`.

**Comment by GeoffreyBooth** [comment]
@benjamn I’ve been meaning to find time to look into this, but I haven’t yet. The only idea I had regarding testing was to do a `$.get` call for the generated file (i.e. the `.js` file output by the `coffeescript` package) and inspecting that file for the source map. Is there some way to know what the URL should be for the `.js` file generated as part of the tests? Or file path, if this is easier to do server-side? (Presumably we don’t need to test for source maps in both places, either it’s in one or it’s in neither.)

**Comment by klaussner** [comment]
What do you think about creating another package, `coffeescript-compiler`, that exports only the CoffeeScript compiler and use that in the `coffeescript` package, similar to `babel-compiler` and `ecmascript`? Then we would have access to code and source maps without running the build plugin.

**Comment by benjamn** [comment]
I would support that split, though it doesn't necessarily need to happen before Meteor 1.4.3, as long the new version of `coffeescript` (depending on `coffeescript-compiler`) is patch-compatible with the current version.

Questions of versioning are about to become somewhat more important, since 1.4.3 will be going back to constraining core packages to be patch-compatible with the version that was current when the Meteor release was published:
* https://github.com/meteor/meteor/pull/8123/commits/868a52829fd26433baf3bde72edbd944cf02e61b
* https://github.com/meteor/meteor/pull/8123/commits/ff4b00c774ef4cbef9ddefea514b426d6694cf90

Given this change, it might make sense to move `packages/coffeescript` to `packages/non-core/coffeescript`, so that it won't be constrained by the release (similar to how it works now, post-1.4 version unpinning).

**Comment by benjamn** [comment]
Another way of asking the same question: would `coffeescript` benefit in any way from being constrained by the Meteor release? Or would you rather be able to bump the version whenever and however you like? If the second option is more appealing, then we should probably do the move to `packages/non-core/coffescript` prior to 1.4.3.

**Comment by GeoffreyBooth** [comment]
I guess the question is, do any Meteor core packages depend on `coffeescript`? I assume not?

Something else to keep in mind is that CoffeeScript 2.0.0-alpha1 is imminent, and will be a breaking-change release of CoffeeScript. The [breaking changes are minimal](https://rawgit.com/jashkenas/coffeescript/2/docs/v2/index.html#breaking-changes), but could affect people, especially people that use a lot of the advanced features of CoffeeScript classes. This might be a reason to let people control what version of CoffeeScript to install into a project.

**Comment by benjamn** [comment]
It sounds like the balance of arguments favors keeping `coffeescript` independent from the release (i.e., moving it into `packages/non-core`). Given that, I'm going to merge this PR into the `release-1.4.3` branch, then `git mv coffeescript non-core/` so that Meteor 1.4.3 won't constrain the `coffeescript` version (and thus won't prevent upgrading to 2.0.0).

**Comment by GeoffreyBooth** [comment]
Thanks @benjamn

**Comment by benjamn** [comment]
Just published [`coffeescript@1.12.3_1`](https://github.com/meteor/meteor/blob/b6c2991a28b2ac4e635081133360b55c1e4387f6/packages/non-core/coffeescript/package.js#L3) using the latest recommended Meteor version, 1.4.2.7.

I tested source maps manually and they seemed to work!

**Comment by GeoffreyBooth** [comment]
@benjamn I noticed that since moving `coffeescript` to non-core, it’s not included in the tests that run via `./meteor test-packages`. I eventually figured out that it can still be tested via `./meteor test-packages packages/non-core/coffeescript`, but you might want to update the regular `test-packages` command to also test the non-core packages. They may not be restricted to a particular Meteor release version, but they’re “core enough” to be part of the Meteor repo and presumably used by a great many Meteor apps. I would hope you CI server runs these tests as part of its builds.

**Pull request merged**